src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -45,34 +60,166 @@ internal partial class RazorCustomMessageTarget
+        var trimRanges = semanticTokensParams is ProvideSemanticTokensRangesParams;
+
+        if (!trimRanges)
+        {
+            var rangeParams = semanticTokensParams as ProvideSemanticTokensRangeParams ?? throw new ArgumentNullException(nameof(ProvideSemanticTokensRangeParams));
+            if (rangeParams.Range is null)
+            {
+                throw new ArgumentNullException(nameof(rangeParams.Range));
+            }
+
+            var newParams = new SemanticTokensRangeParams
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                PartialResultToken = semanticTokensParams.PartialResultToken,
+                Range = rangeParams!.Range,
+            };
+
+            using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId);
+            var csharpResults = await _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangeParams, VSSemanticTokensResponse>(
+                snapshot.TextBuffer,
+                lspMethodName,
+                languageServerName,
+                newParams,
+                cancellationToken).ConfigureAwait(false);
+
+            var result = csharpResults?.Response;
+            failed = result is null;
+            if (!failed)
+            {
+                response = new ProvideSemanticTokensResponse(result!.Data, semanticTokensParams.RequiredHostDocumentVersion);
+            }
+        }
+        else
+        {
+            var rangesParams = semanticTokensParams as ProvideSemanticTokensRangesParams ?? throw new ArgumentNullException(nameof(ProvideSemanticTokensRangesParams));
+            if (rangesParams.Ranges is null)
+            {
+                throw new ArgumentNullException(nameof(rangesParams.Ranges));
+            }
+
+            // Ensure the C# ranges are sorted
+            Array.Sort(rangesParams!.Ranges, static (r1, r2) => r1.CompareTo(r2));
+            var requestTasks = new List<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>(rangesParams!.Ranges.Length);
+
+            foreach (var range in rangesParams!.Ranges)
+            {
+                var newParams = new SemanticTokensRangeParams
+                {
+                    TextDocument = semanticTokensParams.TextDocument,
+                    PartialResultToken = semanticTokensParams.PartialResultToken,
+                    Range = range,
+                };
+
+                var task = _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangeParams, VSSemanticTokensResponse>(
+                    snapshot.TextBuffer,
+                    lspMethodName,
+                    languageServerName,
+                    newParams,
+                    cancellationToken);
+                requestTasks!.Add(task);
+            }
+
+            using var _ = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId);
```

Comment:
As soon as the task is created for a request, it can start processing, so this really should go beofe the loop starts. Also consider a using block to constrain when the timing ends, as at the moment it is including the stitching together of the results in the time.